### DISCUSSION TITLE
A little confused about Auth

### DISCUSSION THREAD

**Discussion by LethargicDeveloper** [pocketbase/pocketbase#5474] [Q&A]
This is my first time using PocketBase. I'm trying to integrate it into a .net core backend. So my frontend will make all requests through the .net backend and the .net backend will forward any necessary requests to PocketBase rather than the frontend calling PocketBase directly. After reading the docs and some discussion topics here I'm still a little confused about how the auth should work.

In my backend I have a /login route. The login route calls `pocketbase.User.AuthenticateWithPasswordAsync(request.Username!, request.Password!);` and if successful returns a token to the caller. That's all well and good and it works.

Then when the user calls and endpoint requiring auth it goes through my PocketBase AuthenticationHandler. The first thing it does is create a NEW instance of the PocketBase client and call `RefreshAsync`

```
var pocketbase = pocketBaseFactory.CreatePocketBase();
pocketbase.AuthStore.Save(token, null);
await pocketbase.User.RefreshAsync();
```

If that's successful, then we're logged in (or have a valid token) and I setup some claims and stuff. And that all works.

Here are my questions:

1) How do I log a user out? I've seen some discussions on this, but this part still confuses me. It seems like the AuthStore is per user. Am I supposed to have a PocketBase client per user and keep track of them? Is the issued token just always going to be valid (unless they do something like change their password)? Basically, I want to be able to invalid a specific token or all tokens associated with a user. Is this something you can only do with the Golang sdk? I'm a bit confused on how I would accomplish this in C#.

2) If I'm creating a new PocketBase client on every auth request to determine if their token is valid (using RefreshAsync), if the token changes (I assume RefreshAsync could return a new token?) what is the most appropriate way to get that token back to the caller?

**Comment by ganigeorgiev** [comment]
I don't think I can help you as I don't have experience with .Net, but note that PocketBase web APIs are completely stateless, aka. there are no sessions in the traditional sense and a user is considered authenticated only if the submitted request has valid `Authorization:TOKEN` header. That's all.

How and where this token will be stored it is up to the user. The official JS and Dart SDKs keeps track of the token as part of `pb.authStore` (this is a local state and it has no relation to the PocketBase server).

That's also the reason why there is no logout action because there are no sessions and there is nothing to "logout". The issued token is not stored in the database and when discarded from the local state it expires on its own.

> So my frontend will make all requests through the .net backend and the .net backend will forward any necessary requests to PocketBase rather than the frontend calling PocketBase directly. 

I really don't recommend doing this, especially if you are new to web development, because most likely you'll mess up the security. PocketBase is designed to be interact directly. Syncing the auth state between 2 different backends comes with a lot of caveats (see https://github.com/pocketbase/pocketbase/discussions/5313) and personally I would need a very good reason to do it. Syncing/proxying the realtime connection will also introduce additional complexity. Whether you need _" new PocketBase client on every auth request"_ depends on a lot of things including how your .Net server operates. See for example the recommendations for JS SSR in https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration

**Reply by LethargicDeveloper** [reply]
I read over your linked srr integration post.

I'm not using ssr. The frontend is a client-rendered react spa. My api that the frontend calls is a .net web api.

From the document:

Problem 1 - Global/shared PocketBase JS SDK server instance
 I'm not using a global/shared PocketBase instance. The PocketBase client used in the api is scoped to each request.
  
Problem 2 - OAuth2 integration
 I'm not integrating with OAuth.
 
Problem 3 - Mixed client and server access
 All PocketBase access is done in the api, so it's not mixed.

My backend is stateless too, so there really isn't any syncing. Just passing a token back and forth.

That said, what is the appropriate way to handle those scenarios:

* Logout - Are all issued tokens always valid until their expiry or you change your password?
What if you want to invalidate a token early? I suppose on logout I could update the password to the same password and invalidate all tokens. I think I read on here you can do that. But what if I just want to invalidate the current token I'm logged in with?

* Refresh Token - If I'm validating the user's token using `pocketbase.User.RefreshAsync()` and that happens to issue a new token, should that token just be ignored? Should the user be responsible for requesting a new token only on 401? So, there will possible be a bunch of unused but valid tokens created?

Or maybe I'm just thinking too hard about it and don't need to worry about all the valid, unused, tokens that could possibly get created.

**Reply by ganigeorgiev** [reply]
> I'm not using ssr. The frontend is a client-rendered react spa. My api that the frontend calls is a .net web api.

Then this makes even less sense to me and I don't understand what you are trying to do. You don't need PocketBase if you just want to use SQLite.

>  Are all issued tokens always valid until their expiry or you change your password?

Yes. All issued tokens are valid until they expire, the user change their password or the shared secret is changed.

> But what if I just want to invalidate the current token I'm logged in with?

There is no builtin mechanism to do that from the web API at the moment (and I'm not sure if there is need for it because even if we add such endpoint it will probably require some confirmation, aka. it will be practically the same as calling "password reset" with the old password).

> If I'm validating the user's token using pocketbase.User.RefreshAsync() and that happens to issue a new token, should that token just be ignored?

It is up to you. You can simply ignore it if you don't need it but if your API doesn't have any state why do you need to explicitly validate it in the first place when the PocketBase API will do that for you when you send the request to the PocketBase server?

> So, there will possible be a bunch of unused but valid tokens created?

Again, I don't see why this should matter. The token is not stored anywhere, it is just returned in the response. Only the client that requested it has access to it. The performance cost of generating a new token is negligible (based on the [benchmarks](https://github.com/pocketbase/benchmarks) `getOne` and `authRefresh` have similar performance).

**Reply by LethargicDeveloper** [reply]
It's not that I wanted to use SQLite. I don't care much, what db pocketbase uses. It was more sort of the all-in-one nature of it. I _could_ use other things for auth or file storage or what-have-you, and if pocketbase isn't the right fit, I may. I just figured I could use pocketbase's auth to secure my api then I don't need to implement it myself.

The app I'm writing is a bit of an experiment anyway and I wanted to give pocketbase a try. It seems I'm just not using it as intended.
